pennylane/math/sparse_matrix.py

```diff
@@ -0,0 +1,245 @@
+    >>> print(s.data)
+    None
+    >>> print(s.shape)
+    (12, 12)
+
+    Args:
+        arg (tuple[int] or tensor_like): if tuple, this is interpreted as the shape of
+            the sparse matrix; else it is interpreted as a 2-d tensor which is a dense representation
+            of the matrix.
+
+    **Example**
+
+    To mimic ``scipy.sparse`` objects, sparse matrices give acces to lists of row and column indices,
+    as well as the number of non-zero elements:
+
+    >>> s = SparseMatrix(tf.Variable([[3., 0.],[2., 0.]]))
+    >>> print(s.row)
+    [0, 1]
+    >>> print(s.col)
+    [0, 0]
+    >>> print(s.nnz)
+    2
+
+    They support elementary arithmetic such as addition, subtraction and scalar multiplication from both sides:
+
+    >>> s1 = SparseMatrix(torch.tensor([[3., 0.],[2., 0.]]))
+    >>> s2 = SparseMatrix(torch.tensor([[-3., 0.],[0., 0.]]))
+    >>> res = s1 + s2
+    >>> print(res.data)
+    {(1, 0): tensor(2.)}
+
+    >>> res2 = s1 - s2
+    >>> print(res2.data)
+    {(0, 0): tensor(6.), (1, 0): tensor(2.)}
+
+    >>> res3 = 2. * s1
+    >>> print(res3.data)
+    {(0, 0): tensor(6.), (1, 0): tensor(4.)}
+
+    We can also compute the Kronecker product (or tensor product) of two sparse matrices:
+
+    >>> res4 = s1.kron(s2)
+    >>> print(res4.data)
+    {(0, 0): tensor(-9.0), (2, 0): tensor(-6.0)}
+    >>> print(res4.shape)
+    (4, 4)
+    """
+
+    def __init__(self, arg):
+
+        if isinstance(arg, tuple):
+            self._data = {}
+            self._shape = arg
+            if len(self._shape) != 2:
+                raise ValueError(f"Expected a 2-dimensional tensor; got shape {self._shape}.")
+
+        else:
+            self._shape = qml.math.shape(arg)
+            if len(self._shape) != 2:
+                raise ValueError(f"Expected a 2-dimensional tensor; got shape {self._shape}.")
+
+            # TODO: rewrite the loop to only loop
+            # over non-zero entries of the input array
+            self._data = {}
+            for i in range(self._shape[0]):
+                for j in range(self._shape[1]):
+                    entry = arg[i, j]
+                    # easiest way to check that entry is not zero
+                    if qml.math.count_nonzero(entry) > 0:
+                        self._data[(i, j)] = entry
+
+    @property
+    def data(self):
+        """dict: return the data of the representation"""
+        return self._data
+
+    @property
+    def shape(self):
+        """tuple: return the shape of the matrix"""
+        return self._shape
+
+    @property
+    def row(self):
+        """list[int]: list of row indices of nonzero entries"""
+        return [idx[0] for idx in self.data]
+
+    @property
+    def col(self):
+        """list[int]: list of column indices of nonzero entries"""
+        return [idx[1] for idx in self.data]
+
+    @property
+    def nnz(self):
+        """int: number of nonzero entries"""
+        return len(self._data)
+
+    def __eq__(self, other):
```

Comment:
How the UI looks like for this? Does it work as `a == b`?

Reply:
Yes exactly!